sdks/python/apache_beam/io/fileio.py

```diff
@@ -34,17 +33,301 @@
+    def equiv_autos(lhs, rhs):
+      return ((lhs == 'AUTO' and rhs == CompressionTypes.AUTO) or
+              (lhs == CompressionTypes.AUTO and rhs == 'AUTO'))
+
+    return (self.file_path == other.file_path and
+            self.start_offset == other.start_offset and
+            self.end_offset == other.end_offset and
+            (self.compression_type == other.compression_type or
+             equiv_autos(self.compression_type, other.compression_type)) and
+            self.coder == other.coder and self.mime_type == other.mime_type)
+
+  @property
+  def path(self):
+    return self.file_path
+
+  def reader(self):
+    return NativeFileSourceReader(self)
+
+
+class NativeFileSourceReader(iobase.NativeSourceReader,
+                             coders.observable.ObservableMixin):
+  """The source reader for a NativeFileSource.
+
+  This class is to be only inherited by source readers natively implemented by
+  Cloud Dataflow service, hence should not be sub-classed by users.
+  """
+
+  def __init__(self, source):
+    super(NativeFileSourceReader, self).__init__()
+    self.source = source
+    self.start_offset = self.source.start_offset or 0
+    self.end_offset = self.source.end_offset
+    self.current_offset = self.start_offset
+
+  def __enter__(self):
+    self.file = ChannelFactory.open(
+        self.source.file_path,
+        'rb',
+        mime_type=self.source.mime_type,
+        compression_type=self.source.compression_type)
+
+    # Determine the real end_offset.
+    #
+    # If not specified or if the source is not splittable it will be the length
+    # of the file (or infinity for compressed files) as appropriate.
+    if ChannelFactory.is_compressed(self.file):
+      if not isinstance(self.source, TextFileSource):
+        raise ValueError('Unexpected compressed file for a non-TextFileSource.')
+
+      self.end_offset = range_trackers.OffsetRangeTracker.OFFSET_INFINITY
+
+    elif self.end_offset is None:
+      self.file.seek(0, os.SEEK_END)
+      self.end_offset = self.file.tell()
+      self.file.seek(self.start_offset)
+
+    # Initializing range tracker after self.end_offset is finalized.
+    self.range_tracker = range_trackers.OffsetRangeTracker(self.start_offset,
+                                                           self.end_offset)
+
+    # Position to the appropriate start_offset.
+    if self.start_offset > 0:
+      if ChannelFactory.is_compressed(self.file):
+        # TODO: Turns this warning into an exception soon.
+        logging.warning(
+            'Encountered initial split starting at (%s) for compressed source.',
+            self.start_offset)
+    self.seek_to_true_start_offset()
+
+    return self
+
+  def __exit__(self, exception_type, exception_value, traceback):
+    self.file.close()
+
+  def __iter__(self):
+    if self.current_offset > 0 and ChannelFactory.is_compressed(self.file):
+      # TODO: Turns this warning into an exception soon.
```

Comment:
Let's add a comment here with something like.

When compression is enabled both initial and dynamic splitting should be prevented. Here we prevent initial splitting by ignoring all splits other than the split that starts at byte 0.